client.go

```diff
@@ -206,130 +243,143 @@ func (c *httpClient) doRequest(ctx context.Context, method, url string, params R
+			log.Printf("[INFO] Received error %v - Retrying", err)
 			return true
 		}
 	}
 	return false
 }
 
-// nolint:unused
-func (c *httpClient) createRequest(ctx context.Context, method, url string, params RequestParams) (*http.Request, *bytes.Buffer, error) {
+func (c *Client) createRequest(ctx context.Context, method, endpoint string, params requestParams) (*http.Request, *bytes.Buffer, error) {
 	var bodyReader io.Reader
 	var bodyBuffer *bytes.Buffer
 
 	if params.Body != nil {
-		bodyBuffer = new(bytes.Buffer)
-		if err := json.NewEncoder(bodyBuffer).Encode(params.Body); err != nil {
-			if c.debug && c.logger != nil {
-				c.logger.Errorf("failed to encode body: %v", err)
-			}
-			return nil, nil, fmt.Errorf("failed to encode body: %w", err)
+		// Reset the body position to the start before using it
+		_, err := params.Body.Seek(0, io.SeekStart)
+		if err != nil {
+			return nil, nil, c.ErrorAndLogf("failed to seek to the start of the body: %v", err.Error())
 		}
-		bodyReader = bodyBuffer
+
+		bodyReader = params.Body
 	}
 
-	req, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
+	req, err := http.NewRequestWithContext(ctx, method, fmt.Sprintf("%s/%s", strings.TrimRight(c.hostURL, "/"),
+		strings.TrimLeft(endpoint, "/")), bodyReader)
 	if err != nil {
-		if c.debug && c.logger != nil {
-			c.logger.Errorf("failed to create request: %v", err)
-		}
-		return nil, nil, fmt.Errorf("failed to create request: %w", err)
+		return nil, nil, c.ErrorAndLogf("failed to create request: %v", err.Error())
 	}
 
+	// Set the default headers
 	req.Header.Set("Content-Type", "application/json")
 	req.Header.Set("Accept", "application/json")
 	if c.userAgent != "" {
 		req.Header.Set("User-Agent", c.userAgent)
 	}
 
+	// Set additional headers added to the client
+	for name, values := range c.header {
+		for _, value := range values {
+			req.Header.Set(name, value)
+		}
+	}
+
 	return req, bodyBuffer, nil
 }
 
-// nolint:unused
-func (c *httpClient) applyBeforeRequest(req *http.Request) error {
+func (c *Client) applyBeforeRequest(req *http.Request) error {
 	for _, mutate := range c.onBeforeRequest {
 		if err := mutate(req); err != nil {
-			if c.debug && c.logger != nil {
-				c.logger.Errorf("failed to mutate before request: %v", err)
-			}
-			return fmt.Errorf("failed to mutate before request: %w", err)
+			e := c.ErrorAndLogf("failed to mutate before request: %v", err.Error())
+			return e
 		}
 	}
+
 	return nil
 }
 
-// nolint:unused
-func (c *httpClient) applyAfterResponse(resp *http.Response) error {
+func (c *Client) applyAfterResponse(resp *http.Response) error {
 	for _, mutate := range c.onAfterResponse {
 		if err := mutate(resp); err != nil {
-			if c.debug && c.logger != nil {
-				c.logger.Errorf("failed to mutate after response: %v", err)
-			}
-			return fmt.Errorf("failed to mutate after response: %w", err)
+			e := c.ErrorAndLogf("failed to mutate after response: %v", err.Error())
+			return e
```

Comment:
Similar to above, can be reduced to one line

Reply:
Fixed!